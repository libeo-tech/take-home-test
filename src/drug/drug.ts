export interface DrugInterface {
  name: string;
  expiresIn: number;
  benefit: number;
  checkBenefit: () => void;
  canUpdateBenefit: () => boolean;
}

export class Drug implements DrugInterface {
  name: string;
  expiresIn: number;
  benefit: number;
  constructor(name, expiresIn, benefit) {
    this.name = name;
    this.expiresIn = expiresIn;
    this.benefit = benefit;
  }

  checkBenefit() {
    if (this.benefit > 50) this.benefit = 50;
    if (this.benefit < 0) this.benefit = 0;
  }

  /* cela marche dans le cadre de l'exercice, 
  cependant ce n'est pas la meilleur faÃ§on de faire si de nouveaux cas particuliers arrivent */
  canUpdateBenefit() {
    return (
      (this.benefit < 50 && this.benefit > 0) ||
      (this.name === FERVEX && this.benefit !== 0)
    );
  }
}

export const DEFAULT = "default";
export const MAGIC_PILL = "Magic Pill";
export const HERBAL_TEA = "Herbal Tea";
const FERVEX = "Fervex";
const DAFALGAN = "Dafalgan";

const BENEFIT_CHANGE = 1;

type updateBenefitFunction = (benefit: number, expiresIn?: number) => number;

const getDefaultBenefitUpdate = (expiresIn: number): number =>
  expiresIn > 0 ? BENEFIT_CHANGE : BENEFIT_CHANGE * 2;

const updateDefautDrugBenefit: updateBenefitFunction = (benefit, expiresIn) =>
  benefit - getDefaultBenefitUpdate(expiresIn);

const updateUpgradingDrugsBenefit: updateBenefitFunction = (
  benefit,
  expiresIn
): number => benefit + getDefaultBenefitUpdate(expiresIn);

const upgradeDafalganBenefit: updateBenefitFunction = (benefit, expiresIn) =>
  benefit - 2 * getDefaultBenefitUpdate(expiresIn);

const updateFervexBenefit: updateBenefitFunction = (benefit, expiresIn) => {
  if (expiresIn <= 0) return 0;
  if (expiresIn <= 5) return benefit + BENEFIT_CHANGE * 3;
  if (expiresIn <= 10) return benefit + BENEFIT_CHANGE * 2;
  return benefit + BENEFIT_CHANGE;
};

export const benefitEvolutionRules: { [key: string]: updateBenefitFunction } = {
  [DEFAULT]: updateDefautDrugBenefit,
  [HERBAL_TEA]: updateUpgradingDrugsBenefit,
  [FERVEX]: updateFervexBenefit,
  [MAGIC_PILL]: benefit => benefit,
  [DAFALGAN]: upgradeDafalganBenefit
};
