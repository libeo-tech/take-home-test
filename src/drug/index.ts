const HERBAL_TEA = "Herbal Tea";
const MAGIC_PILL = "Magic Pill";
const FERVEX = "Fervex";
const DAFALGAN = "Dafalgan";

const DEFAULT_BENEFIT_DECREASE = 1;

export class Drug {
  name: string;
  expiresIn: number;
  benefit: number;

  constructor(name: string, expiresIn: number, benefit: number) {
    this.name = name;
    this.expiresIn = expiresIn;
    this.benefit = benefit;
  }

  decreaseExpiration() {
    if (this.name !== MAGIC_PILL) this.expiresIn--;
  }

  applyBenefitConstraints(): void {
    if (this.benefit > 50) this.benefit = 50;
    else if (this.benefit < 0) this.benefit = 0;
  }

  hasBenefitLimitBeenReached(): boolean {
    return !(this.benefit > 0 && this.benefit < 50);
  }

  computeNextBenefit(): void {
    if (this.hasBenefitLimitBeenReached()) return;

    switch (this.name) {
      case MAGIC_PILL:
        break;
      case HERBAL_TEA:
        this.benefit += degradeBenefit(this.expiresIn);
        break;

      case DAFALGAN:
        this.benefit -= 2 * degradeBenefit(this.expiresIn);
        break;

      case FERVEX:
        if (this.expiresIn < 1) this.benefit = 0;
        else if (this.expiresIn < 6)
          this.benefit += DEFAULT_BENEFIT_DECREASE * 3;
        else if (this.expiresIn < 11)
          this.benefit += DEFAULT_BENEFIT_DECREASE * 2;
        else this.benefit += DEFAULT_BENEFIT_DECREASE;

        break;

      default:
        this.benefit -= degradeBenefit(this.expiresIn);
        break;
    }

    this.applyBenefitConstraints();
  }
}

const degradeBenefit = (expiresIn: number): number =>
  expiresIn > 0 ? DEFAULT_BENEFIT_DECREASE : DEFAULT_BENEFIT_DECREASE * 2;
