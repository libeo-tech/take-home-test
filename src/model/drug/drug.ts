import { boundNumber } from '../../tools/boundNumber';
import { DrugName } from '../../type/DrugName';

export abstract class Drug {
  private _internalBenefit: number;

  protected constructor(
    public readonly name: DrugName,
    public expiresIn: number,
    benefit: number
  ) {
    this.benefit = benefit;
  }

  public get benefit(): number {
    return this._internalBenefit;
  }

  /**
   * Makes benefit >= 0 before setting it
   * @param value
   */
  public set benefit(value: number) {
    this._internalBenefit = boundNumber(value, 0, 50);
  }

  /**
   * Returns true if expiresIn <= 0
   */
  public isExpired() {
    return this.expiresIn <= 0;
  }

  /**
   * Update internal expiration and benefit values according to drug's rules
   */
  public updateValues(): this {
    this.benefit -= this.isExpired() ? 2 : 1;
    this.expiresIn -= 1;

    return this;
  }

  /**
   * Ensures output consistency when using JSON.stringify()
   */
  public toJSON() {
    return {
      name: this.name,
      expiresIn: this.expiresIn,
      benefit: this.benefit,
    };
  }

  /**
   * Ensures output consistency when using Drug.toString()
   */
  public toString(): string {
    return JSON.stringify(this);
  }
}
