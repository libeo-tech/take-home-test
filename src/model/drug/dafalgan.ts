import { DrugName } from '../../type/DrugName';
import { Drug } from './drug';

export class Dafalgan extends Drug {
  constructor(expiresIn: number, benefit: number) {
    super(DrugName.DAFALGAN, expiresIn, benefit);
  }

  public updateValues(): this {
    this.benefit -= this.isExpired() ? 4 : 2;
    this.expiresIn -= 1;

    return this;
  }
}
