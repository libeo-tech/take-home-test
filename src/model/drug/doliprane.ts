import { DrugName } from '../../type/DrugName';
import { Drug } from './drug';

export class Doliprane extends Drug {
  constructor(expiresIn: number, benefit: number) {
    super(DrugName.DOLIPRANE, expiresIn, benefit);
  }
}
