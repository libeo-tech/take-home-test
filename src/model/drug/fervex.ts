import { DrugName } from '../../type/DrugName';
import { Drug } from './drug';

export class Fervex extends Drug {
  constructor(expiresIn: number, benefit: number) {
    super(DrugName.FERVEX, expiresIn, benefit);
  }

  public updateValues() {
    if (this.expiresIn > 10) {
      this.benefit += 1;
    } else if (this.expiresIn > 5) {
      this.benefit += 2;
    } else if (this.expiresIn > 0) {
      this.benefit += 3;
    } else {
      this.benefit = 0;
    }
    this.expiresIn -= 1;

    return this;
  }
}
