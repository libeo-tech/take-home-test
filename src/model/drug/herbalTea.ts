import { DrugName } from '../../type/DrugName';
import { Drug } from './drug';

export class HerbalTea extends Drug {
  constructor(expiresIn: number, benefit: number) {
    super(DrugName.HERBAL_TEA, expiresIn, benefit);
  }

  public updateValues() {
    if (this.isExpired()) {
      this.benefit += 2;
    } else {
      this.benefit += 1;
    }
    this.expiresIn -= 1;

    return this;
  }
}
