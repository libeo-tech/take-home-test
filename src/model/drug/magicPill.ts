import { DrugName } from '../../type/DrugName';
import { Drug } from './drug';

export class MagicPill extends Drug {
  constructor(expiresIn: number, benefit: number) {
    super(DrugName.MAGIC_PILL, expiresIn, benefit);
  }

  public isExpired(): boolean {
    return false;
  }

  public updateValues() {
    return this;
  }
}
