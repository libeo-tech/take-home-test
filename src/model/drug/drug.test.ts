import { DrugName } from '../../type/DrugName';
import { Drug } from './drug';

class TestDrug extends Drug {
  constructor(expiresIn: number, benefit: number) {
    super(DrugName.TEST, expiresIn, benefit);
  }
}

describe('Drug', () => {
  let drug: Drug;

  beforeEach(() => {
    drug = new TestDrug(30, 5);
  });

  it('should not allow for a negative benefit', () => {
    drug.benefit = -1;
    expect(drug).toMatchSnapshot();
  });

  it('should not allow for a benefit greater than 50', () => {
    drug.benefit = 51;
    expect(drug).toMatchSnapshot();
  });

  it('should return expiration status', () => {
    expect(drug.isExpired()).toBe(false);
    drug.expiresIn = 0;
    expect(drug.isExpired()).toBe(true);
  });

  it('should decrease expiresIn and benefit over time', () => {
    drug.updateValues();
    expect(drug).toMatchSnapshot();
  });

  it('should decrease benefit twice as much if expired', () => {
    drug.expiresIn = 0;
    drug.updateValues();
    expect(drug).toMatchSnapshot();
  });

  it('should be serializable to JSON', () => {
    expect(drug.toJSON()).toMatchSnapshot();
  });

  it('should be serializable to string', () => {
    expect(drug.toString()).toMatchSnapshot();
  });
});
