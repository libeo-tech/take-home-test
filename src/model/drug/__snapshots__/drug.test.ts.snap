// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drug should be serializable to JSON 1`] = `
Object {
  "benefit": 5,
  "expiresIn": 30,
  "name": "test",
}
`;

exports[`Drug should be serializable to string 1`] = `"{\\"name\\":\\"test\\",\\"expiresIn\\":30,\\"benefit\\":5}"`;

exports[`Drug should decrease benefit twice as much if expired 1`] = `
Object {
  "benefit": 3,
  "expiresIn": -1,
  "name": "test",
}
`;

exports[`Drug should decrease expiresIn and benefit over time 1`] = `
Object {
  "benefit": 4,
  "expiresIn": 29,
  "name": "test",
}
`;

exports[`Drug should not allow for a benefit greater than 50 1`] = `
Object {
  "benefit": 50,
  "expiresIn": 30,
  "name": "test",
}
`;

exports[`Drug should not allow for a negative benefit 1`] = `
Object {
  "benefit": 0,
  "expiresIn": 30,
  "name": "test",
}
`;
